Compression/src/CompressDetStepMCs_module.cc

```diff
@@ -0,0 +1,638 @@
+// Note that the number of detector steps is NOT reduced.
+//
+// Allowed compression levels for different data products:
+// - StrawGasSteps : noCompression
+// - CaloShowerSteps : noCompression
+// - CrvSteps : noCompression
+// - SimParticles : noCompression, fullCompression
+// - StepPointMCs : noCompression, simParticleCompression
+// - MCTrajectories : noCompression, simParticleCompression
+//
+// There is also the concept of "genealogy" compression, which uses the fhicl parameter
+// keepNGenerations and takes an integer corresponding to the
+// number of generations back you want to keep. All missing generations are
+// replaced with a new SimParticle to identify that a truncation has occured
+// - Note 1: N = -1 means keep all generations (i.e. no compression)
+// - Note 2: the very first SimParticle (i.e. the one that has a valid Ptr to a GenParticle) is always kept
+//
+// Dec 2020, Andy Edmonds
+//
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "art_root_io/TFileService.h"
+
+#include <memory>
+
+#include "MCDataProducts/inc/StrawGasStep.hh"
+#include "MCDataProducts/inc/CaloShowerStep.hh"
+#include "MCDataProducts/inc/CrvStep.hh"
+#include "MCDataProducts/inc/StepPointMCCollection.hh"
+#include "MCDataProducts/inc/SimParticleCollection.hh"
+#include "Mu2eUtilities/inc/compressSimParticleCollection.hh"
+#include "MCDataProducts/inc/GenParticleCollection.hh"
+#include "Compression/inc/CompressionLevel.hh"
+#include "MCDataProducts/inc/MCTrajectoryCollection.hh"
+#include "MCDataProducts/inc/MCRelationship.hh"
+
+namespace mu2e {
+  class CompressDetStepMCs;
+
+  typedef std::set<art::Ptr<SimParticle> > SimParticleSet;
+
+  class SimParticleSelector {
+  public:
+    SimParticleSelector(const SimParticleSet& simPartSet) {
+      for (const auto& i_simPart : simPartSet) {
+        cet::map_vector_key key = cet::map_vector_key(i_simPart.key());
+        m_keys.insert(key);
+      }
+    }
+
+    bool operator[]( cet::map_vector_key key ) const {
+      return m_keys.find(key) != m_keys.end();
+    }
+
+    const std::set<cet::map_vector_key>& keys() const {
+      return m_keys;
+    }
+
+    void clear() {
+      m_keys.clear();
+    }
+
+  private:
+    std::set<cet::map_vector_key> m_keys;
+
+  };
+
+  typedef std::string InstanceLabel;
+}
+
+
+class mu2e::CompressDetStepMCs : public art::EDProducer {
+public:
+  struct Config {
+    using Name=fhicl::Name;
+    using Comment=fhicl::Comment;
+
+    // fhicl parameters for all of our inputs
+    fhicl::Atom<art::InputTag> strawGasStepTag{Name("strawGasStepTag"), Comment("InputTag for the StrawGasSteps")};
```

Comment:
For mixing we will have multiple Step collections.  I think the step input config needs to be a Sequence, not an Atom.

Reply:
I don't see us running this module when mixing unless I've missed something. Is the idea to create datasets with primary+pileup steps? I thought we would be taking primary steps and pileup steps to create primary+pileup digis.